In [70]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
# https://github.com/DataTalksClub/nyc-tlc-data/releases/tag/green/download
for i in range(3): 
    flink = f"https://github.com/DataTalksClub/nyc-tlc-data/releases/download/green/green_tripdata_2020-1{i}.csv.gz"
    print(f"#{i}: {flink}")
    !wget {flink}
!ls green_tripdata_2020*

#0: https://github.com/DataTalksClub/nyc-tlc-data/releases/download/green/green_tripdata_2020-10.csv.gz
--2024-02-09 01:25:30--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/green/green_tripdata_2020-10.csv.gz
Распознаётся github.com (github.com)… 140.82.121.3
Подключение к github.com (github.com)|140.82.121.3|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 302 Found
Адрес: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/b94c0bd2-e629-406d-aecf-d4e2c0412807?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240208%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240208T212531Z&X-Amz-Expires=300&X-Amz-Signature=25d97f22ee55ba07409ce81ae39fbbe4b509c84f8e8921e39aa607f9cb3ddbdb&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dgreen_tripdata_2020-10.csv.gz&response-content-type=application%2Foctet-stream [переход]
--2024-

In [71]:
# https://chat.openai.com/share/7b445074-5c5f-4d0f-80a5-57edf93bf9b5
import pandas as pd

# List of file names
file_names = ['green_tripdata_2020-10.csv.gz', 'green_tripdata_2020-11.csv.gz', 'green_tripdata_2020-12.csv.gz']

# Read each file into a separate DataFrame
dfs = [pd.read_csv(file) for file in file_names]

# Concatenate the DataFrames into a single DataFrame
combined_df = pd.concat(dfs, ignore_index=True)

combined_df.head(3)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-10-01 00:31:19   2020-10-01 00:34:55                  N   
1       2.0  2020-10-01 00:42:12   2020-10-01 00:43:51                  N   
2       2.0  2020-10-01 00:53:09   2020-10-01 00:55:39                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0             7             7              1.0           0.79   
1         1.0           179             7              1.0           0.50   
2         1.0           179           223              1.0           0.60   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          5.0    0.5      0.5        1.58           0.0        NaN   
1          4.0    0.5      0.5        0.00           0.0        NaN   
2          4.0    0.5      0.5        1.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          7.88           1.0        1.0   
1                    0.3          5.30           2.0        1.0   
2                    0.3          6.36           1.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0

Add a transformer block and perform the following:

Remove rows where the passenger count is equal to 0 and the trip distance is equal to zero.

Create a new column lpep_pickup_date by converting lpep_pickup_datetime to a date.

Rename columns in Camel Case to Snake Case, e.g. VendorID to vendor_id.

Add three assertions:
- vendor_id is one of the existing values in the column (currently)
- passenger_count is greater than 0
- trip_distance is greater than 0

https://chat.openai.com/share/7b445074-5c5f-4d0f-80a5-57edf93bf9b5

In [13]:
import pandas as pd

# List of file names
file_names = ['green_tripdata_2020-10.csv.gz', 'green_tripdata_2020-11.csv.gz', 'green_tripdata_2020-12.csv.gz']

# Read each file into a separate DataFrame
dfs = [pd.read_csv(file, low_memory=False) for file in file_names]

# Concatenate the DataFrames into a single DataFrame
combined_df = pd.concat(dfs, ignore_index=True)
combined_df.head(3)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-10-01 00:31:19   2020-10-01 00:34:55                  N   
1       2.0  2020-10-01 00:42:12   2020-10-01 00:43:51                  N   
2       2.0  2020-10-01 00:53:09   2020-10-01 00:55:39                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0             7             7              1.0           0.79   
1         1.0           179             7              1.0           0.50   
2         1.0           179           223              1.0           0.60   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          5.0    0.5      0.5        1.58           0.0        NaN   
1          4.0    0.5      0.5        0.00           0.0        NaN   
2          4.0    0.5      0.5        1.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          7.88           1.0        1.0   
1                    0.3          5.30           2.0        1.0   
2                    0.3          6.36           1.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0

In [72]:
# Transformer Block
def transform_data(df):
    # Remove rows where passenger count is equal to 0 and trip distance is equal to zero
    df = df[(df['passenger_count'] > 0) & (df['trip_distance'] > 0)]

    # Create a new column lpep_pickup_date by converting lpep_pickup_datetime to a date
    df['lpep_pickup_date'] = pd.to_datetime(df['lpep_pickup_datetime']).dt.date

    # Rename columns in Camel Case to Snake Case
    df.columns = [col.lower() if col == 'VendorID' else col.lower().replace(' ', '_') for col in df.columns]

    return df

# Apply the transformer block to the combined DataFrame
combined_df = transform_data(combined_df)
combined_df.head(3)

vendorid lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-10-01 00:31:19   2020-10-01 00:34:55                  N   
1       2.0  2020-10-01 00:42:12   2020-10-01 00:43:51                  N   
2       2.0  2020-10-01 00:53:09   2020-10-01 00:55:39                  N   

   ratecodeid  pulocationid  dolocationid  passenger_count  trip_distance  \
0         1.0             7             7              1.0           0.79   
1         1.0           179             7              1.0           0.50   
2         1.0           179           223              1.0           0.60   

   fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          5.0  ...      0.5        1.58           0.0        NaN   
1          4.0  ...      0.5        0.00           0.0        NaN   
2          4.0  ...      0.5        1.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          7.88           1.0        1.0   
1                    0.3          5.30           2.0        1.0   
2                    0.3          6.36           1.0        1.0   

   congestion_surcharge  lpep_pickup_date  
0                   0.0        2020-10-01  
1                   0.0        2020-10-01  
2                   0.0        2020-10-01  

[3 rows x 21 columns]

In [74]:
# # Assertions
# assert combined_df['vendor_id'].isin([1, 2]).all(), "VendorID is not one of the existing values in the column."
# assert (combined_df['passenger_count'] > 0).all(), "Passenger count is not greater than 0."
# assert (combined_df['trip_distance'] > 0).all(), "Trip distance is not greater than 0."

# combined_df.head(3)

### https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/cohorts/2024/02-workflow-orchestration/homework.md#assignment

Question 1. Data Loading
Once the dataset is loaded, what's the shape of the data?
- 266,855 rows x 20 columns
- 544,898 rows x 18 columns
- 544,898 rows x 20 columns
- 133,744 rows x 20 columns

In [77]:
len(combined_df.columns), combined_df['vendorid'].count()

(21, 139370)

### https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/cohorts/2024/02-workflow-orchestration/homework.md#question-2-data-transformation
Question 2. Data Transformation
Upon filtering the dataset where the passenger count is greater than 0 and the trip distance is greater than zero, how many rows are left?

- 544,897 rows
- 266,855 rows
- 139,370 rows
- 266,856 rows

In [80]:
df = combined_df
df = df[(df['passenger_count'] > 0) & (df['trip_distance'] > 0)]
df['vendorid'].count()

139370

### Question 3. Data Transformation - https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/cohorts/2024/02-workflow-orchestration/homework.md#question-3-data-transformation
Which of the following creates a new column lpep_pickup_date by converting lpep_pickup_datetime to a date?

- data = data['lpep_pickup_datetime'].date
- data('lpep_pickup_date') = data['lpep_pickup_datetime'].date
- data['lpep_pickup_date'] = data['lpep_pickup_datetime'].dt.date
- data['lpep_pickup_date'] = data['lpep_pickup_datetime'].dt().date()

In [90]:
file = 'green_tripdata_2020-12.csv.gz'
df = pd.read_csv(file, low_memory=False)
data = df
data.head(2)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-12-01 00:29:37   2020-12-01 00:32:51                  N   
1       2.0  2020-12-01 00:41:46   2020-12-01 00:46:31                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            75            75              1.0           0.59   
1         1.0            75            74              1.0           1.24   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          4.5    0.5      0.5        1.16           0.0        NaN   
1          6.0    0.5      0.5        1.46           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          6.96           1.0        1.0   
1                    0.3          8.76           1.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0

In [91]:
data['lpep_pickup_datetime'] = pd.to_datetime(data['lpep_pickup_datetime'])  # AttributeError: 'Series' object has no attribute 'date'
# data('lpep_pickup_date') = data['lpep_pickup_datetime'].date  # SyntaxError: cannot assign to function call here. Maybe you meant '==' instead of '='?
# data['lpep_pickup_date'] = data['lpep_pickup_datetime'].dt.date  # AttributeError: Can only use .dt accessor with datetimelike values
# data['lpep_pickup_date'] = data['lpep_pickup_datetime'].dt().date()  # AttributeError: Can only use .dt accessor with datetimelike values
data.head(3)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-12-01 00:29:37   2020-12-01 00:32:51                  N   
1       2.0  2020-12-01 00:41:46   2020-12-01 00:46:31                  N   
2       2.0  2020-12-01 00:05:46   2020-12-01 00:10:48                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            75            75              1.0           0.59   
1         1.0            75            74              1.0           1.24   
2         1.0           244           243              2.0           1.19   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          4.5    0.5      0.5        1.16           0.0        NaN   
1          6.0    0.5      0.5        1.46           0.0        NaN   
2          6.0    0.5      0.5        1.46           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          6.96           1.0        1.0   
1                    0.3          8.76           1.0        1.0   
2                    0.3          8.76           1.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0

In [92]:
data['lpep_pickup_date'] = data['lpep_pickup_datetime'].dt.date
data.head(3)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-12-01 00:29:37   2020-12-01 00:32:51                  N   
1       2.0  2020-12-01 00:41:46   2020-12-01 00:46:31                  N   
2       2.0  2020-12-01 00:05:46   2020-12-01 00:10:48                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            75            75              1.0           0.59   
1         1.0            75            74              1.0           1.24   
2         1.0           244           243              2.0           1.19   

   fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          4.5  ...      0.5        1.16           0.0        NaN   
1          6.0  ...      0.5        1.46           0.0        NaN   
2          6.0  ...      0.5        1.46           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          6.96           1.0        1.0   
1                    0.3          8.76           1.0        1.0   
2                    0.3          8.76           1.0        1.0   

   congestion_surcharge  lpep_pickup_date  
0                   0.0        2020-12-01  
1                   0.0        2020-12-01  
2                   0.0        2020-12-01  

[3 rows x 21 columns]

### Question 4. Data Transformation 
- https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/cohorts/2024/02-workflow-orchestration/homework.md#question-4-data-transformation

What are the existing values of VendorID in the dataset?
- 1, 2, or 3
- 1 or 2
- 1, 2, 3, 4
- 1

In [69]:
file_names = ['green_tripdata_2020-10.csv.gz', 'green_tripdata_2020-11.csv.gz', 'green_tripdata_2020-12.csv.gz']
dfs = [pd.read_csv(file) for file in file_names]
df = pd.concat(dfs, ignore_index=True)

df['vendor_id'] = df['VendorID'].fillna(-1).astype(int)

display(
    df['vendor_id'].unique()
)

array([ 2,  1, -1])

Question 5. Data Transformation
How many columns need to be renamed to snake case?
- 3
- 6
- 2
- 4

In [62]:
a = [c for c in df.columns if '_' not in c and c.lower() != c]
len(a), a

(4, ['VendorID', 'RatecodeID', 'PULocationID', 'DOLocationID'])

In [58]:
import pandas as pd

# Assuming combined_df is your DataFrame
# You can replace it with the actual name of your DataFrame

# Convert 'lpep_pickup_datetime' to datetime type
combined_df['lpep_pickup_datetime'] = pd.to_datetime(combined_df['lpep_pickup_datetime'])

# Extract the date part and create a new column 'lpep_pickup_date'
combined_df['lpep_pickup_date'] = combined_df['lpep_pickup_datetime'].dt.date

# Drop the original 'lpep_pickup_datetime' column if needed
# combined_df = combined_df.drop(columns=['lpep_pickup_datetime'])

# Display the DataFrame with the new 'lpep_pickup_date' column
combined_df.head(3)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2020-10-01 00:31:19   2020-10-01 00:34:55                  N   
1       2.0  2020-10-01 00:42:12   2020-10-01 00:43:51                  N   
2       2.0  2020-10-01 00:53:09   2020-10-01 00:55:39                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0             7             7              1.0           0.79   
1         1.0           179             7              1.0           0.50   
2         1.0           179           223              1.0           0.60   

   fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          5.0  ...      0.5        1.58           0.0        NaN   
1          4.0  ...      0.5        0.00           0.0        NaN   
2          4.0  ...      0.5        1.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          7.88           1.0        1.0   
1                    0.3          5.30           2.0        1.0   
2                    0.3          6.36           1.0        1.0   

   congestion_surcharge  lpep_pickup_date  
0                   0.0        2020-10-01  
1                   0.0        2020-10-01  
2                   0.0        2020-10-01  

[3 rows x 21 columns]

In [93]:
distinct_date_count = combined_df['lpep_pickup_date'].nunique()
print("Distinct Count of lpep_pickup_date:", distinct_date_count)

Distinct Count of lpep_pickup_date: 95


Question 6. Data Exporting
Once exported, how many partitions (folders) are present in Google Cloud?
- 96
- 56
- 67
- 108